File: bookstore-ts/tests/util/pagination.test.ts
import { expect, test, describe } from "bun:test";
import { paginate } from "../../src/util/pagination";

describe("paginate", () => {
  const cases: { name: string; page: number; size: number; wantLimit: number; wantOffset: number }[] = [
    { name: "first page", page: 1, size: 10, wantLimit: 10, wantOffset: 0 },
    { name: "second page", page: 2, size: 10, wantLimit: 10, wantOffset: 10 },
    { name: "third page custom size", page: 3, size: 5, wantLimit: 5, wantOffset: 10 },
    { name: "large page", page: 100, size: 20, wantLimit: 20, wantOffset: 1980 },
  ];
  for (const tt of cases) {
    test(tt.name, () => {
      const [limit, offset] = paginate(tt.page, tt.size);
      expect(limit).toBe(tt.wantLimit);
      expect(offset).toBe(tt.wantOffset);
    });
  }
});

File: bookstore-go/internal/util/pagination_test.go
package util_test

import (
	"testing"

	"github.com/xebia/ai-assisted-se-claude-exercise/internal/util"
)

func TestPaginate(t *testing.T) {
	tests := []struct {
		name       string
		page, size int
		wantLimit  int
		wantOffset int
	}{
		{"first page", 1, 10, 10, 0},
		{"second page", 2, 10, 10, 10},
		{"third page custom size", 3, 5, 5, 10},
		{"large page", 100, 20, 20, 1980},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			limit, offset := util.Paginate(tt.page, tt.size)
			if limit != tt.wantLimit || offset != tt.wantOffset {
				t.Errorf("Paginate(%d, %d) = (%d, %d), want (%d, %d)",
					tt.page, tt.size, limit, offset, tt.wantLimit, tt.wantOffset)
			}
		})
	}
}
